narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
     Round,
 };
 
 #[cfg(test)]
 #[path = "tests/certificate_waiter_tests.rs"]
 pub mod certificate_waiter_tests;
 
-/// The resolution of the timer that checks whether we received replies to our parent requests, and triggers
-/// a round of GC if we didn't.
-const GC_RESOLUTION: u64 = 10_000;
+// Maximum number of certficates to fetch with one request.
+const MAX_CERTIFICATES_TO_FETCH: usize = 1000;
 
-/// Waits to receive all the ancestors of a certificate before looping it back to the `Core`
-/// for further processing.
+/// When there are certificates missing from local store, e.g. discovered when a received certificate has missing parents,
+/// CertificateWaiter is responsible for fetching missing certificates from other primaries.
 pub struct CertificateWaiter {
+    /// Identity of the current authority.
+    name: PublicKey,
     /// The committee information.
     committee: Committee,
+    /// Network client to fetch certificates from other primaries.
+    network: Arc<dyn PrimaryToPrimaryRpc>,
     /// The persistent storage.
-    store: CertificateStore,
-    /// Receiver for signal of round change
-    rx_consensus_round_updates: watch::Receiver<u64>,
-    /// The depth of the garbage collector.
-    gc_depth: Round,
+    certificate_store: CertificateStore,
     /// Watch channel notifying of epoch changes, it is only used for cleanup.
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
-    /// Receives sync commands from the `Synchronizer`.
-    rx_synchronizer: Receiver<Certificate>,
-    /// Loops back to the core certificates for which we got all parents.
-    tx_core: Sender<Certificate>,
-    /// List of digests (certificates) that are waiting to be processed. Their processing will
-    /// resume when we get all their dependencies. The map holds a cancellation `Sender`
-    /// which we can use to give up on a certificate.
-    // TODO: remove the OnceCell once drain_filter stabilizes
-    pending: DashMap<HeaderDigest, (Round, OnceCell<oneshot::Sender<()>>)>,
+    /// Receives certificates with missing parents from the `Synchronizer`.
+    rx_certificate_waiter: Receiver<Certificate>,
+    /// Kicks start a certificate fetching task, after receiving a certificate with missing parents.
+    /// Sending to this channel should use `try_send()`, because only one pending message is needed.
+    tx_kick_from_cert: mpsc::Sender<()>,
+    rx_kick_from_cert: mpsc::Receiver<()>,
+    /// Kicks start a certificate fetching task, after finishing the current fetch task.
+    /// Sending to this channel should use `send()`. When a fetch task is finishing, the channel should be empty.
+    /// The purpose is to ensure a new fetch task is started after the previous fetch task ends.
+    tx_kick_from_fetch: mpsc::Sender<()>,
+    rx_kick_from_fetch: mpsc::Receiver<()>,
+    /// Loops fetched certificates back to the core. Certificates are ensured to have all parents.
```

Comment:
asking for my own future planning, is this loopback something you think I could replace with a request-response model from core->certificate_waiter? Or is there a reason that would not be possible?

Reply:
I think sending the fetched certificates from certificate waiter to core via network seems inefficient. And I added a oneshot::channel for core to signal back to certificate waiter that it has finished processing the looped back certs. If there is a way to make this communication a direct function call, that would be nice. Otherwise I don't know another way to improve on the loopback channel for now.

Reply:
Yeah I didn't mean using the network! I meant more like making it a direct function call. Something I will keep in mind for the future but not related to this PR now, thanks :)